crates/sats/src/relation.rs

```diff
@@ -255,34 +384,77 @@ impl Header {
+        &self.fields.columns
+    }
+
+    /// Returns the [Column] at the specified position.
+    pub fn column_by_pos(&self, pos: usize) -> &Column {
+        &self.fields[pos]
     }
 
-    pub fn column_pos<'a>(&'a self, col: &'a FieldName) -> Option<ColId> {
+    /// Returns the [FieldName] at the specified position.
+    pub fn field_by_pos(&self, pos: usize) -> Option<&FieldName> {
+        self.fields.column_by_pos(pos).map(|x| x.field.as_ref())
+    }
+
+    /// Returns the [Column] with the specified [FieldName].
+    pub fn column(&self, col: &FieldName) -> Option<&Column> {
         match col {
-            FieldName::Name { .. } => self.fields.iter().position(|f| &f.field == col),
-            FieldName::Pos { field, .. } => self
-                .fields
-                .iter()
-                .enumerate()
-                .position(|(pos, f)| &f.field == col || *field == pos),
+            FieldName::Name { .. } => self.fields.get_field(col),
+            FieldName::Pos { field, .. } => self.fields.column_by_pos(*field),
         }
-        .map(Into::into)
     }
 
-    pub fn column_pos_or_err<'a>(&'a self, col: &'a FieldName) -> Result<ColId, RelationError> {
-        self.column_pos(col)
+    /// Returns the [Column] with the specified `field_name`.
+    pub fn column_by_name(&self, field_name: &str) -> Option<&Column> {
+        self.column(&FieldName::named(&self.table_name, field_name))
+    }
+
+    /// Returns the [FieldName] with the specified `field_name`.
+    pub fn field_by_name(&self, field_name: &str) -> Option<&FieldName> {
+        self.column_by_name(field_name).map(|x| x.field.as_ref())
+    }
+
+    /// Returns the [Column] with the specified [ColId].
+    pub fn column_by_id(&self, col_id: ColId) -> Option<&Column> {
+        self.fields.iter().find(|x| x.col_id == col_id)
+    }
+
+    /// Returns the [FieldName] with the specified [ColId].
+    pub fn field_by_id(&self, col_id: ColId) -> Option<&FieldName> {
+        self.column_by_id(col_id).map(|x| x.field.as_ref())
+    }
+
+    /// Returns the [ColId] of the specified [FieldName].
+    pub fn col_id_by_field<'a>(&'a self, col: &'a FieldName) -> Option<ColId> {
+        self.column(col).map(|x| x.col_id)
+    }
+
+    /// Returns the [ColId] of the specified [FieldName] or [RelationError::FieldNotFound].
+    pub fn col_id_by_field_or_err<'a>(&'a self, col: &'a FieldName) -> Result<ColId, RelationError> {
+        self.col_id_by_field(col)
             .ok_or_else(|| RelationError::FieldNotFound(self.clone_for_error(), col.clone()))
     }
 
-    /// Finds the position of a field with `name`.
-    pub fn find_pos_by_name(&self, name: &str) -> Option<ColId> {
-        self.column_pos(&FieldName::named(&self.table_name, name))
+    /// Finds the [ColId] of a field with `name`.
+    pub fn find_col_id_by_name(&self, name: &str) -> Option<ColId> {
+        self.col_id_by_field(&FieldName::named(&self.table_name, name))
+    }
```

Comment:
This still has an unneeded allocation. That's because `FieldName::named` takes a `&str` and creates an owned `String` from it. But we're just doing a lookup into a map. So we shouldn't need to allocate at all.

So we really need to update `col_id_by_field` to take a `&str` instead.